multidimensional-pod-autoscaler/KEP.md

```diff
@@ -0,0 +1,499 @@
+
+**MPA Updater.** MPA Updater will update the number of replicas in the deployment and evict the eligible pods for vertical scaling.
+
+**MPA Admission-Controller.** If users intend to directly execute the autoscaling recommendations generated from the MPA Recommender, the MPA Admission-Controller will update the deployment configuration (i.e., the size of each replica) and configure the rolling update to the Application Deployment.
+
+### Action Actuation Implementation
+
+To actuate the decisions without losing availability, we plan to (1) evict pods with min-replicas configured and update Pod sizes with the web-hooked admission controller (for vertical scaling), and (2) add or remove replicas (for horizontal scaling).
+We use a web-hooked admission controller to manage vertical scaling because if the actuator directly updates the vertical scaling configurations through deployment, it will potentially overload etcd (as vertical scaling might be quite frequent).
+MPA Admission Controller intercepts Pod creation requests and rewrites the request by applying recommended resources to the Pod spec.
+We do not use the web-hooked admission controller to manage the horizontal scaling as it could slow down the pod creation process.
+In the future when the in-place vertical resizing is enabled, we can remove the web-hooked admission controller and only have the updater.
+
+[<img src="./kep-imgs/mpa-action-actuation.png" width="400"/>](./kep-imgs/mpa-action-actuation.png "MPA Action Actuation")
+
+Pros:
+- Vertical scaling is handled by webhooks to avoid overloading etcd
+- Horizontal scaling is handled through deployment to avoid extra overhead by webhooks
+- Authentication and authorization for vertical scaling are handled by admission webhooks
+- Recommendation and the actuation are completely separated
+
+Cons:
+- Webhooks introduce extra overhead for vertical scaling operations (can be avoided after in-place resizing of pod is enabled without eviction)
+- Vertical and horizontal scaling executions are separated (can be avoided after in-place resizing of pod is enabled without eviction)
+- State changes in pod sizes are not persisted (too much to keep in etcd, could use Prometheus to store pod state changes)
+
+### Action Recommendation Implementation
+
+To generate the vertical scaling action recommendation, we reuse VPA libraries as much as possible to implement scaling algorithm integrated with the newly generated MPA API code.
+To do that, we need to update accordingly the code which reads and updates the VPA objects to be interacting with the MPA objects.
+To generate the horizontal scaling action recommendation, we reuse HPA libraries, integrating with the MPA API code, to reads and updates the MPA objects.
```

Comment:
Are we certain the existing HPA code can be reused as a library, or will this potentially require changes in k/k to make this code more easily reusable? Is there a risk that we become dependent on guarantees that the HPA k/k code doesn't provide?

Reply:
There is no need to change code k/k for now.

Reply:
I'm not sure here. I believe it was possible to import HPA code and reuse it with little modification. But HPA code in k/k doesn't seem to have a well defined API to be used as library at the moment. Let's think about the right API for code reuse in HPA so that we don't end up with having to fix MPA whenever current method signatures change in HPA. I guess we would need concrete PR to see if the current API is fine or what is missing.

Interesting side-note: how will we know when making code changes to k/k HPA that we don't break MPA by mistake?

Reply:
That's a good point. The current plan is to have a maintainer watch the imported HPA library for incorporating new changes that will necessarily affect MPA.